src/restrictededitingmodeediting.js

```diff
@@ -284,3 +304,60 @@ function filterDeleteCommandsOnMarkerBoundaries( selection, markerRange ) {
 		return true;
 	};
 }
+
+// Ensures that model.deleteContent() does not delete outside exception markers ranges.
+//
+// The enforced restrictions are:
+// - only execute deleteContent() inside exception markers
+// - restrict passed selection to exception marker
+function restrictDeleteContent( editor ) {
+	return ( evt, args ) => {
+		const [ selection ] = args;
+
+		const marker = getMarkerAtPosition( editor, selection.focus ) || getMarkerAtPosition( editor, selection.anchor );
+
+		// Stop method execution if marker was not found at selection focus.
+		if ( !marker ) {
+			evt.stop();
+
+			return;
+		}
+
+		// Collapsed selection inside exception marker does not require fixing.
+		if ( selection.isCollapsed ) {
+			return;
+		}
+
+		const allowedToDelete = marker.getRange().getIntersection( selection.getFirstRange() );
+
+		// Shrink the selection to the range inside exception marker.
+		selection.setTo( allowedToDelete );
```

Comment:
I don't think that you should be modifying the selection passed to the command. Instead, override `args[ 0 ]` with a new selection instance. 

Also, `DocumentSelection` doesn't even have `setTo()`. Please add a test for that – I think it would crash.